nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/nar/NarRestApiClient.java

```diff
@@ -0,0 +1,138 @@
+
+    private final URI baseUri;
+    private final WebClientService webClientService;
+    private final ObjectMapper objectMapper;
+
+    public NarRestApiClient(final WebClientService webClientService, final String host, final int port, final boolean secure) {
+        try {
+            this.baseUri = new URI(secure ? HTTPS_SCHEME : HTTP_SCHEME, null, host, port, null, null, null);
+        } catch (URISyntaxException e) {
+            throw new RuntimeException(e);
+        }
+        this.webClientService = Objects.requireNonNull(webClientService, "WebClientService is required");
+        this.objectMapper = new ObjectMapper();
+        this.objectMapper.setAnnotationIntrospector(new JakartaXmlBindAnnotationIntrospector(objectMapper.getTypeFactory()));
+    }
+
+    public NarSummariesEntity listNarSummaries() {
+        final URI requestUri = new StandardHttpUriBuilder()
+                .scheme(baseUri.getScheme())
+                .host(baseUri.getHost())
+                .port(baseUri.getPort())
+                .addPathSegment(NIFI_API_PATH)
+                .addPathSegment(CONTROLLER_PATH)
+                .addPathSegment(NAR_MANAGER_PATH)
+                .addPathSegment(NARS_PATH)
+                .build();
+        LOGGER.info("Requesting NAR summaries from {}", requestUri);
+
+        final HttpRequestBodySpec requestBodySpec = webClientService.get()
+                .uri(requestUri)
+                .header("Accept", "application/json");
+
+        try (final HttpResponseEntity response = requestBodySpec.retrieve()) {
+            final InputStream responseBody = getResponseBody(requestUri, response);
+            return objectMapper.readValue(responseBody, NarSummariesEntity.class);
+        } catch (final WebClientServiceException | IOException e) {
+            throw new NarRestApiRetryableException(e.getMessage(), e);
+        }
+    }
+
+    public InputStream downloadNar(final String identifier) {
+        final URI requestUri = new StandardHttpUriBuilder()
+                .scheme(baseUri.getScheme())
+                .host(baseUri.getHost())
+                .port(baseUri.getPort())
+                .addPathSegment(NIFI_API_PATH)
+                .addPathSegment(CONTROLLER_PATH)
+                .addPathSegment(NAR_MANAGER_PATH)
+                .addPathSegment(NARS_PATH)
+                .addPathSegment(identifier)
+                .addPathSegment(NAR_DOWNLOAD_PATH)
+                .build();
+        LOGGER.info("Downloading NAR [{}] from {}", identifier, requestUri);
+
+        try {
+            final HttpResponseEntity response = webClientService.get()
+                    .uri(requestUri)
+                    .header("Accept", "application/octet-stream")
+                    .retrieve();
+            return getResponseBody(requestUri, response);
+        } catch (final WebClientServiceException e) {
+            throw new NarRestApiRetryableException(e.getMessage(), e);
```

Comment:
Recommend providing the identifier in the message:
```suggestion
            throw new NarRestApiRetryableException("Download NAR [%s] failed".formatted(identifier), e);
```

Reply:
Same reasoning as next comment